pinc/project_states.inc

```diff
@@ -40,262 +40,282 @@ class ProjectState
+    {
+        return self::$states[$state]->medium_label ?? '';
+    }
 
-// Note that the order in which these project states are declared
-// is the order in which they will be displayed in various contexts
-// (via $PROJECT_STATES_IN_ORDER).
+    public static function text($state)
+    {
+        return self::$states[$state]->long_label ?? '';
+    }
 
-function declare_project_states($rounds, $pools)
-{
-    global $waiting_projects_forum_idx, $projects_forum_idx, $pp_projects_forum_idx, $posted_projects_forum_idx,
-    $completed_projects_forum_idx, $deleted_projects_forum_idx, $project_states_for_star_metal_;
-    global $PROJECT_STATES_IN_ORDER, $project_states;
-
-    // for the initial creation of a project
-    declare_project_state(
-        "PROJ_NEW",
-        "project_new",
-        _("New Project"),
-        _("New Project"),
-        $waiting_projects_forum_idx,
-        'NEW',
-        ''
-    );
+    public static function get_forum($state)
+    {
+        return self::$states[$state]->forum ?? -1;
+    }
 
-    foreach ($rounds as $round) {
-        // Populate $round->project_states with all project states that apply
-        // to this round for lookup later.
-        $round->project_states[] = $round->project_bad_state = "{$round->id}.proj_bad";
-        $round->project_states[] = $round->project_unavailable_state = "{$round->id}.proj_unavail";
-        $round->project_states[] = $round->project_waiting_state = "{$round->id}.proj_waiting";
-        $round->project_states[] = $round->project_available_state = "{$round->id}.proj_avail";
-        $round->project_states[] = $round->project_complete_state = "{$round->id}.proj_done";
-
-        declare_project_state(
-            "PROJ_{$round->id}_BAD_PROJECT",
-            $round->project_bad_state,
-            "$round->id: " .  _("Bad Project"),
-            "$round->name: " . _("Bad Project"),
-            $projects_forum_idx,
-            'PAGE_EDITING',
-            ''
-        );
+    public static function get_phase($state)
+    {
+        return self::$states[$state]->phase ?? 'NONE';
+    }
 
-        declare_project_state(
-            "PROJ_{$round->id}_UNAVAILABLE",
-            $round->project_unavailable_state,
-            "$round->id: " . _("Unavailable"),
-            "$round->name: " . _("Unavailable"),
-            ($round->id == 'P1' ? $waiting_projects_forum_idx : $projects_forum_idx),
-            'PAGE_EDITING',
+    // Note that the order in which these project states are declared
+    // is the order in which they will be displayed in various contexts
+    // (via ProjectStates::get_states()).
+
+    public static function declare_project_states($rounds, $pools)
+    {
+        global $waiting_projects_forum_idx, $projects_forum_idx, $pp_projects_forum_idx, $posted_projects_forum_idx,
+        $completed_projects_forum_idx, $deleted_projects_forum_idx;
+
+        // for the initial creation of a project
+        self::declare_project_state(
```

Comment:
It would be clearer to define the constant explicitly and add the state rather than have it all inside a function. Maybe

```php
    define("PROJ_NEW", "project_new");
    ProjectStates::add_state(PROJ_NEW, new ProjectState(
        _("New Project"),
        _("New Project"),
        $waiting_projects_forum_idx,
        'NEW',
        ''
    ));
```
Admitedly this involves a repetition but only of a constant or member variable so the compiler would catch any mistakes.